### PR TITLE
🐛 prevent blocking of KCP and DockerMachine controllers

### LINKED ISSUE #6006: KCP InfrastructureMachineTemplate not deleted after Cluster deletion if infrastructure never becomes ready
**What steps did you take and what happened:**

1- Create a Cluster where infrastructure never becomes ready (e.g., AWS does not have enough resources)
2- Delete the Cluster and `InfrastructureMachineTemplate` referenced by KCP never gets deleted.

KCP does not set owner reference to its `InfrastructureMachineTemplate` until infrastructure becomes ready and if the `Cluster` is deleted, `InfrastructureMachineTemplate` is not cleaned up due to the missing owner reference.

This issue is not happening with `MachineDeployments` because the owner reference is set at the reconcile.

**What did you expect to happen:**

**Anything else you would like to add:**
Owner reference of `InfrastructureMachineTemplate` needs to be set before this point: https://github.com/kubernetes-sigs/cluster-api/blob/5d5dc7a1d50842577b32fbc41bcedc48925f125c/controlplane/kubeadm/internal/controllers/controller.go#L153

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

Currently the KCP  controller are blocked if the `InfrastructureCluster` is not ready. This leads to cases where if the infrastructure is never ready because of any reason (Example: resource limit reached in AWS) one cannot even delete the cluster - the operation is blocked at deleting KCP. This PR fixes the reconciler so that we no longer blocking in such a case.

This PR also fixes a similar problem with blocking DockerMachine reconciliation if infrastructure is not ready.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6006

### CODE REVIEW

**Comment by ykakarap** [comment]
Local testing using CAPD seems to be working fine with the new change to KCP reconciler. Will wait for the e2e results to check if anything breaks.

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 1]
```diff
@@ -255,6 +255,12 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, err
 	}
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller.go:257]: Should we drop l.153-156?

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 2]
```diff
@@ -192,6 +185,13 @@ func patchDockerMachine(ctx context.Context, patchHelper *patch.Helper, dockerMa
 func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, dockerMachine *infrav1.DockerMachine, externalMachine *docker.Machine, externalLoadBalancer *docker.LoadBalancer) (res ctrl.Result, retErr error) {
 	log := ctrl.LoggerFrom(ctx)
 
+	// Check if the infrastructure is ready, otherwise return and wait for the cluster object to be updated
```
Comment (sbueringer) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: Should we document this in the migration doc?

Do we have similar code in CABPK or KCP?

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 1 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller.go:257]: Woops, looks like i forgot to delete those lines. Will do.

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go (hunk 2 above)
Comment (ykakarap) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: This PR already addresses similar case in KCP. I will do a pass on all of your controllers to check if we have similar blocks in other controllers.

Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: Ups yeah of course KCP is already addressed :)

Comment (ykakarap) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: >Should we document this in the migration doc?

On further thought,  this behavior is very internal to CAPI and does not affect how providers would use CAPI, correct?
Is the idea to highlight that CAPI does not block delete operations on KCP and DockerMachine if Infrastructure is not ready? (a point in the `other` section of the 1.1->1.2 doc?)

Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: My thought was. If this applies to our infra provider (CAPD), it's probably something that applies to other infra providers too.

Similar for bootstrap & control plane providers.

Comment (vincepri) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: +1 to get some documentation for providers in case this is missed. Instead of the migration doc (or in addition to), we should have this note in the book section for provider implementers

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 3]
```diff
@@ -246,7 +246,7 @@ func patchKubeadmControlPlane(ctx context.Context, patchHelper *patch.Helper, kc
 }
 
 // reconcile handles KubeadmControlPlane reconciliation.
-func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (res ctrl.Result, reterr error) {
+func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (res ctrl.Result, reterr error) { //nolint:gocyclo
```
Comment (killianmuldoon) [inline controlplane/kubeadm/internal/controllers/controller.go:249]: Is this far off passing gocyclo? It would be  better to restructure a little maybe than to add this tag whenever a function goes over the limit.

**Review by killianmuldoon — APPROVED** [review]
lgmt +/- documentation and gocyclo fix.

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 3 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller.go:249]: I think it just passed the limit - basing this on the observation that the gocyclo warning showed up after adding a new if branch to this function.

[the remark this answers is not in the document]
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller.go]: Bumped up gocyclo complexity limit from 30 (default) to 50.

File: .golangci.yml [hunk 4]
```diff
@@ -44,6 +44,8 @@ linters:
   - whitespace
 
 linters-settings:
+  gocyclo:
+    min-complexity: 50
```
[the remark this answers is not in the document]
Comment (vincepri) [reply .golangci.yml:48]: I originally suggested to do this to @ykakarap , although not sure at this point if it makes sense to keep it around; reducing the number of if branches or lines of code in a particular function or method isn't a particularly useful target. We should always strive for readability and simplicity rather than refactoring based on a arbitrary number.

[the remark this answers is not in the document]
Comment (ykakarap) [reply .golangci.yml:48]: >Is the previous level 30? (default from https://golangci-lint.run/usage/linters/#gocyclo)

Yes. 

I am +1 to dropping it. And looks like Vince is also okay with dropping it: https://github.com/kubernetes-sigs/cluster-api/pull/6183#discussion_r812047541

[the remark this answers is not in the document]
Comment (killianmuldoon) [reply .golangci.yml:48]: I'm fine with dropping or upping the limit -  agreed it's not the most useful linter, but it does work as a sort of a low pass filter for complexity. @sbueringer @fabriziopandini  WDYT?

[the remark this answers is not in the document]
Comment (sbueringer) [reply .golangci.yml:48]: I'm fine with dropping it given that our most complex code is not even found by gocyclo and it's mostly finding relatively straightforward code right now.

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go (hunk 2 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: I'm not sure if we currently have documentation on that level e.g. how to implement a infra machine/cluster reconciler (but maybe we have)

Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: I think the most detailed doc is: https://cluster-api.sigs.k8s.io/developer/providers/implementers-guide/controllers_and_reconciliation.html
But that doesn't even come close to just looking at KCP / CABPK / CAPD

I don't think we can keep a detailed documentation reasonably up-to-date over time

Comment (ykakarap) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: This seemed to be the best place to add the note: https://github.com/kubernetes-sigs/cluster-api/pull/6183/files#diff-1b5f4c4ac9a4463df3857e86252b263cabf6207cf4844a9814f20f698b510d2eR116-R117

Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: That link doesn't work for me (GitHub UI just doesn't do anything when I click on it)

Comment (ykakarap) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: https://github.com/ykakarap/cluster-api/blob/kcp-machinetemplate-cleanup/docs/book/src/developer/providers/machine-infrastructure.md?plain=1#L116-L117

File: .golangci.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ykakarap) [reply .golangci.yml:48]: Will drop it.

**Comment by sbueringer** [comment]
I think that's not a flake. Probably the unit test is now running slightly differently and now needs the CRD.

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go (hunk 2 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/controllers/dockermachine_controller.go:188]: Looks like the perfect place, agree

**Comment by ykakarap** [comment]
> I think that's not a flake. Probably the unit test is now running slightly differently and now needs the CRD.

I have seem the test pass a few times and fail a few times in my local. 🤔

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 5]
```diff
@@ -147,7 +147,7 @@ func TestReconcileReturnErrorWhenOwnerClusterIsMissing(t *testing.T) {
 	g.Expect(result).To(Equal(ctrl.Result{}))
 
 	// calling reconcile should return error
-	g.Expect(env.Delete(ctx, cluster)).To(Succeed())
+	g.Expect(env.CleanupAndWait(ctx, cluster)).To(Succeed())
```
Comment (ykakarap) [inline controlplane/kubeadm/internal/controllers/controller_test.go:150]: If the cache is not updated to reflect the cluster deletion then this test panic. Adding a `CleanupAndWait` makes sure that the test tests exactly the desired thing.

File: .golangci.yml [hunk 6]
```diff
@@ -12,7 +12,6 @@ linters:
   - gci
   - goconst
   - gocritic
-  - gocyclo
```
Comment (fabriziopandini) [inline .golangci.yml:15]: This seems unrelated to the goal of this PR 🤔

File: docs/book/src/developer/providers/v1.1-to-v1.2.md [hunk 7]
```diff
@@ -31,4 +31,7 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
 
 ### Other
 
--
+- KubeadmControlPlane controller and DockerMachineMachine controller no longer block if `cluster.status.infrastructureReady` is false.
+This fixes the following problems:
+  - Cluster deletion leaves behind orphaned machine templates referenced in KCP.
+  - CAPD cluster deletion is blocked if infrastructure is not ready.
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:37]: I think that we should use this section to document only what the providers are expected to do or things that can impact them while upgrading (it is not required to document all the changes applied to KCP/CAPD if they do not impact providers)

what about rephrasing in.

```suggestion
after investigating the issue... we discovered that improper implementation of a check on cluster.status.infrastructureReady can lead to problems during cluster deletion.

as a consequence, we recommend all the providers check that
- the check for cluster.status.infrastructureReady=true usually existing at the beginning of the reconcile loop for control-plane providers is implemented after setting external objects ref and 
- the check for cluster.status.infrastructureReady=true usually existing  at the beginning of the reconcile loop for infrastructure provider does not prevent the object to be deleted
```

**Comment by ykakarap** [comment]
/test pull-cluster-api-test-main

running again to test for flakes

File: .golangci.yml (hunk 6 above)
Comment (ykakarap) [reply .golangci.yml:15]: This PR will hit the gocyclo limit on the kcp controllers reconcile function. This was done to address that. Ref discussion here: https://github.com/kubernetes-sigs/cluster-api/pull/6183#discussion_r812134821

Shall I drop the gocyclo linter in a separate PR and then rebase this one?

File: controlplane/kubeadm/internal/controllers/controller_test.go (hunk 5 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller_test.go:150]: Okay now I understand why the test panic'ed and I also understand why that fixes the panic in the test.

I assume this panic can't happen in reality because the code panic'ed because `r.managementCluster` was nil?

Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller_test.go:150]: For this particular test case it panics at the `List` operation of the `GetCRDWithContract` function because the client used here is the `APIReader` which is not provided in the test case. Even if we add the `APIReader` in the test case the test would still panic because `r.managementCluster` is nil.

File: .golangci.yml (hunk 6 above)
Comment (vincepri) [reply .golangci.yml:15]: Separate PR sounds good to me

Comment (ykakarap) [reply .golangci.yml:15]: PR: https://github.com/kubernetes-sigs/cluster-api/pull/6193

File: controlplane/kubeadm/internal/controllers/controller_test.go (hunk 5 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller_test.go:150]: Okay got it. So overall this panic only happens in tests. Thx

File: docs/book/src/developer/providers/v1.1-to-v1.2.md [hunk 8]
```diff
@@ -31,4 +31,6 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
 
 ### Other
 
--
+- After the investigating the [issue](https://github.com/kubernetes-sigs/cluster-api/issues/6006) we discovered that improper implementation of a check on `cluster.status.infrastructureReady` can lead to problems during cluster deletion. As a consequence, we recommend all the providers check that:
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:34]:
```suggestion
- After the investigating the [issue](https://github.com/kubernetes-sigs/cluster-api/issues/6006) we discovered that improper implementation of a check on `cluster.status.infrastructureReady` can lead to problems during cluster deletion. As a consequence, we recommend all the providers to ensure that:
```

File: docs/book/src/developer/providers/v1.1-to-v1.2.md [hunk 9]
```diff
@@ -31,4 +31,6 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
...
+  - the check for `cluster.status.infrastructureReady=true` usually existing at the beginning of the reconcile loop for control-plane providers is implemented after setting external objects ref and
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:35]:
```suggestion
  - The check for `cluster.status.infrastructureReady=true` usually existing at the beginning of the reconcile loop for control-plane providers is implemented after setting external objects ref;
```

File: docs/book/src/developer/providers/v1.1-to-v1.2.md [hunk 10]
```diff
@@ -31,4 +31,6 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
...
+  - the check for `cluster.status.infrastructureReady=true` usually existing  at the beginning of the reconcile loop for infrastructure provider does not prevent the object to be deleted
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:36]:
```suggestion
  - The check for `cluster.status.infrastructureReady=true` usually existing  at the beginning of the reconcile loop for infrastructure provider does not prevent the object to be deleted
  
rif. https://github.com/kubernetes-sigs/cluster-api/pull/6183
```

Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:36]:
```suggestion
  - The check for `cluster.status.infrastructureReady=true` usually existing  at the beginning of the reconcile loop for infrastructure provider does not prevent the object to be deleted
  
rif. https://github.com/kubernetes-sigs/cluster-api/pull/6183
```

**Review by fabriziopandini — COMMENTED** [review]
nits

**Comment by ykakarap** [comment]
/test pull-cluster-api-test-mink8s-main


test-infra flake.

File: docs/book/src/developer/providers/v1.1-to-v1.2.md [hunk 11]
```diff
@@ -74,3 +74,8 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
 
       This change has been introduced in CAPI in the following PRs: [#6072](https://github.com/kubernetes-sigs/cluster-api/pull/6072), [#6190](https://github.com/kubernetes-sigs/cluster-api/pull/6190).</br>
       **Note**: This change is not mandatory for providers, but highly recommended.
+
+- After the investigating the [issue](https://github.com/kubernetes-sigs/cluster-api/issues/6006) we discovered that improper implementation of a check on `cluster.status.infrastructureReady` can lead to problems during cluster deletion. As a consequence, we recommend all the providers to ensure that:
```
Comment (sbueringer) [inline docs/book/src/developer/providers/v1.1-to-v1.2.md:78]:
```suggestion
- After investigating an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/6006) we discovered that improper implementation of a check on `cluster.status.infrastructureReady` can lead to problems during cluster deletion. As a consequence, we recommend that all providers ensure that:
```
nit + rebase

**Comment by sbueringer** [comment]
Now only rebase :)

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-full-main

/hold
let's please wait for the results of full main

**Comment by sbueringer** [comment]
Interesting.
Test fails with:
> /home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/node_drain_timeout.go:82
> Timed out after 540.000s.
> Error: Unexpected non-nil/non-zero argument at index 1:
> 	<*errors.errorString>: &errors.errorString{s:"operator \"Exists\" cannot be converted into the old label selector 
 format"}
> /home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/controlplane_helpers.go:422

Makes kind of sense with that selector: `cluster.x-k8s.io/cluster-name=node-drain-fbrumk,cluster.x-k8s.io/control-plane`
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6183/pull-cluster-api-e2e-full-main/1500855121049817088/artifacts/clusters/bootstrap/resources/node-drain-qb28o3/KubeadmControlPlane/node-drain-fbrumk-control-plane.yaml

Just wonder why this didn't happen before.

/test pull-cluster-api-e2e-full-main

**Comment by sbueringer** [comment]
> @sbueringer the e2e full pipeline is green. Do we still need the hold?
Nope

/hold cancel
/lgtm

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve
